File: 513/main.py
"""
 Given a binary tree, find the leftmost value in the last row of the tree.

Example 1:

Input:

    2
   / \
  1   3

Output:
1

Example 2:

Input:

        1
       / \
      2   3
     /   / \
    4   5   6
       /
      7

Output:
7

Note: You may assume the tree (i.e., the given root node) is not NULL.
"""
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def findBottomLeftValue(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        return self.findBottomLeftValueRecur([root])

    def findBottomLeftValueRecur(self, row):
        next = []
        for node in row:
            if node.left != None:
                next.append(node.left)
            if node.right != None:
                next.append(node.right)
        if len(next) == 0:
            return row[0].val
        else:
            return self.findBottomLeftValueRecur(next)

ans = Solution()

File: 513/main.go
package main

import "fmt"

func main() {
	tr := &TreeNode{
		1,
		&TreeNode{
			2,
			&TreeNode{4, nil, nil},
			nil,
		},
		&TreeNode{
			3,
			&TreeNode{
				5,
				&TreeNode{7, nil, nil},
				nil,
			},
			&TreeNode{6, nil, nil},
		},
	}
	fmt.Println(findBottomLeftValue(tr))
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func findBottomLeftValue(root *TreeNode) int {
	queue := []*TreeNode{root}
	last := 0
	for len(queue) != 0 {
		last = queue[0].Val
		newQueue := []*TreeNode{}
		for _, node := range queue {
			if node.Left != nil {
				newQueue = append(newQueue, node.Left)
			}
			if node.Right != nil {
				newQueue = append(newQueue, node.Right)
			}
		}
		queue = newQueue
	}
	return last
}
